File: legacy/old_package_copy/SARdine/python/sardine/quality_flags.py
"""
Quality Flags for SARdine Processing
====================================

This module implements the quality flag system required by the scientific audit
(NOISE-1, DEBURST-1, RTC-1). Quality flags are collected during processing and
embedded in output metadata.

Usage:
    from sardine.quality_flags import QualityFlagCollector, QualityFlag

    # Create collector at start of processing
    collector = QualityFlagCollector()
    
    # Add flags during processing
    if noise_skipped:
        collector.add_flag(QualityFlag.NOISE_REMOVAL_SKIPPED, 
                          reason="Zero ratio exceeded threshold",
                          zero_ratio=0.96, threshold=0.95)
    
    # Get flags for metadata
    flags_dict = collector.to_dict()
"""

import os
import logging
from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from enum import Enum, auto
from typing import Any, Dict, List, Optional, Set
import json

logger = logging.getLogger(__name__)


class QualityFlag(Enum):
    """Quality flag types for scientific processing validation."""
    
    # NOISE-1: Noise removal conditions
    NOISE_REMOVAL_SKIPPED = auto()
    NOISE_THRESHOLDS_MODIFIED = auto()
    
    # DEBURST-1: Deburst fallback conditions  
    DEBURST_T0_FALLBACK = auto()
    DEBURST_TIME_DOMAIN_MISMATCH = auto()
    
    # RTC-1: Terrain correction conditions
    RTC_DEFAULT_INCIDENCE_ANGLE = auto()
    RTC_INCIDENCE_PROFILE_MISSING = auto()
    
    # Calibration conditions
    CALIBRATION_LUT_WARNING = auto()
    
    # General
    WARNING = auto()
    ERROR = auto()


# Severity mapping
FLAG_SEVERITY: Dict[QualityFlag, str] = {
    QualityFlag.NOISE_REMOVAL_SKIPPED: "WARNING",
    QualityFlag.NOISE_THRESHOLDS_MODIFIED: "INFO",
    QualityFlag.DEBURST_T0_FALLBACK: "WARNING",
    QualityFlag.DEBURST_TIME_DOMAIN_MISMATCH: "ERROR",
    QualityFlag.RTC_DEFAULT_INCIDENCE_ANGLE: "WARNING",
    QualityFlag.RTC_INCIDENCE_PROFILE_MISSING: "WARNING",
    QualityFlag.CALIBRATION_LUT_WARNING: "WARNING",
    QualityFlag.WARNING: "WARNING",
    QualityFlag.ERROR: "ERROR",
}


@dataclass
class QualityFlagEntry:
    """Individual quality flag entry with context."""
    flag_type: str
    severity: str
    message: str
    timestamp: str
    details: Dict[str, Any] = field(default_factory=dict)
    
    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


class QualityFlagCollector:
    """
    Collects quality flags during processing.
    
    Thread-safe collection of quality flags that can be embedded in output metadata.
    """
    
    def __init__(self):
        self._flags: List[QualityFlagEntry] = []
        self._flag_types_seen: Set[str] = set()
    
    def add_flag(
        self,
        flag: QualityFlag,
        message: str = "",
        **details: Any
    ) -> None:
        """
        Add a quality flag.
        
        Args:
            flag: The QualityFlag type
            message: Human-readable description
            **details: Additional context (e.g., threshold values, ratios)
        """
        entry = QualityFlagEntry(
            flag_type=flag.name,
            severity=FLAG_SEVERITY.get(flag, "WARNING"),
            message=message,
            timestamp=datetime.now(timezone.utc).isoformat(),
            details=details,
        )
        
        self._flags.append(entry)
        self._flag_types_seen.add(flag.name)
        
        # Also log the flag
        log_msg = f"🚩 Quality flag: {flag.name}"
        if message:
            log_msg += f" - {message}"
        if details:
            log_msg += f" ({details})"
            
        if entry.severity == "ERROR":
            logger.error(log_msg)
        elif entry.severity == "WARNING":
            logger.warning(log_msg)
        else:
            logger.info(log_msg)
    
    def has_flag(self, flag: QualityFlag) -> bool:
        """Check if a specific flag type is present."""
        return flag.name in self._flag_types_seen
    
    def has_errors(self) -> bool:
        """Check if any error-level flags are present."""
        return any(f.severity == "ERROR" for f in self._flags)
    
    def has_warnings(self) -> bool:
        """Check if any warning-level flags are present."""
        return any(f.severity == "WARNING" for f in self._flags)
    
    def count(self) -> int:
        """Get total number of flags."""
        return len(self._flags)
    
    def get_flags(self) -> List[QualityFlagEntry]:
        """Get all flag entries."""
        return list(self._flags)
    
    def get_flags_by_severity(self, severity: str) -> List[QualityFlagEntry]:
        """Get flags filtered by severity."""
        return [f for f in self._flags if f.severity == severity]
    
    def to_dict(self) -> Dict[str, Any]:
        """
        Convert to dictionary for embedding in metadata.
        
        Returns a dict with:
        - flags: List of all flag entries
        - summary: Dict mapping flag types to counts
        - has_errors: Boolean
        - has_warnings: Boolean
        """
        summary: Dict[str, int] = {}
        for f in self._flags:
            summary[f.flag_type] = summary.get(f.flag_type, 0) + 1
        
        return {
            "flags": [f.to_dict() for f in self._flags],
            "summary": summary,
            "has_errors": self.has_errors(),
            "has_warnings": self.has_warnings(),
            "total_count": len(self._flags),
        }
    
    def to_json(self, indent: int = 2) -> str:
        """Convert to JSON string."""
        return json.dumps(self.to_dict(), indent=indent)
    
    def clear(self) -> None:
        """Clear all flags."""
        self._flags.clear()
        self._flag_types_seen.clear()


def check_noise_thresholds() -> Optional[QualityFlagEntry]:
    """
    Check if noise thresholds are modified by environment variables.
    
    NOISE-1: This surfaces when env vars are modifying noise removal behavior.
    
    Returns:
        QualityFlagEntry if thresholds are modified, None otherwise.
    """
    default_zero = 0.95
    default_ratio = 0.9
    
    zero_threshold = os.environ.get("SARDINE_NOISE_ZERO_THRESHOLD")
    ratio_threshold = os.environ.get("SARDINE_NOISE_RATIO_THRESHOLD")
    
    if zero_threshold is not None or ratio_threshold is not None:
        try:
            actual_zero = float(zero_threshold) if zero_threshold else default_zero
            actual_ratio = float(ratio_threshold) if ratio_threshold else default_ratio
        except ValueError:
            actual_zero = default_zero
            actual_ratio = default_ratio
        
        return QualityFlagEntry(
            flag_type=QualityFlag.NOISE_THRESHOLDS_MODIFIED.name,
            severity="INFO",
            message=f"Noise thresholds modified by environment variables",
            timestamp=datetime.now(timezone.utc).isoformat(),
            details={
                "zero_threshold": actual_zero,
                "ratio_threshold": actual_ratio,
                "default_zero_threshold": default_zero,
                "default_ratio_threshold": default_ratio,
                "SARDINE_NOISE_ZERO_THRESHOLD": zero_threshold,
                "SARDINE_NOISE_RATIO_THRESHOLD": ratio_threshold,
            }
        )
    
    return None


def check_rtc_incidence_available(incidence_near: Optional[float], incidence_far: Optional[float]) -> Optional[QualityFlagEntry]:
    """
    Check if RTC incidence angle is available from annotation.
    
    RTC-1: This surfaces when default incidence angle would be used.
    
    Args:
        incidence_near: Near-range incidence angle from annotation
        incidence_far: Far-range incidence angle from annotation
        
    Returns:
        QualityFlagEntry if incidence angles are missing, None otherwise.
    """
    if incidence_near is None and incidence_far is None:
        return QualityFlagEntry(
            flag_type=QualityFlag.RTC_DEFAULT_INCIDENCE_ANGLE.name,
            severity="WARNING",
            message="Using default incidence angle (35°) - annotation values not available",
            timestamp=datetime.now(timezone.utc).isoformat(),
            details={
                "default_angle_deg": 35.0,
                "reason": "Incidence angle not available from annotation metadata",
            }
        )
    
    return None


# Global collector for the current processing run
_current_collector: Optional[QualityFlagCollector] = None


def get_collector() -> QualityFlagCollector:
    """Get or create the current quality flag collector."""
    global _current_collector
    if _current_collector is None:
        _current_collector = QualityFlagCollector()
    return _current_collector


def reset_collector() -> QualityFlagCollector:
    """Reset and return a new quality flag collector."""
    global _current_collector
    _current_collector = QualityFlagCollector()
    return _current_collector


if __name__ == "__main__":
    # Self-test
    collector = QualityFlagCollector()
    
    # Test adding flags
    collector.add_flag(
        QualityFlag.NOISE_REMOVAL_SKIPPED,
        message="Zero ratio exceeded threshold",
        zero_ratio=0.96,
        threshold=0.95
    )
    
    collector.add_flag(
        QualityFlag.DEBURST_T0_FALLBACK,
        message="Missing polynomial t0 for burst 3",
        burst_id=3
    )
    
    # Test checks
    flag = check_noise_thresholds()
    if flag:
        collector._flags.append(flag)
    
    print("=== Quality Flags ===")
    print(collector.to_json())
    
    print(f"\nHas errors: {collector.has_errors()}")
    print(f"Has warnings: {collector.has_warnings()}")
    print(f"Total count: {collector.count()}")

File: legacy/old_package_copy/SARdine/src/core/quality_flags.rs
//! Quality flags for scientific processing validation.
//!
//! This module implements the quality flag system required by the scientific audit
//! (NOISE-1, DEBURST-1, RTC-1). Quality flags are surfaced in output metadata
//! instead of silent fallbacks.
//!
//! # Usage
//!
//! ```rust,ignore
//! use sardine::core::quality_flags::{QualityFlags, QualityFlag};
//!
//! let mut flags = QualityFlags::new();
//! flags.add(QualityFlag::NoiseRemovalSkipped {
//!     reason: "Zero ratio exceeded threshold".to_string(),
//!     zero_ratio: 0.96,
//!     threshold: 0.95,
//! });
//!
//! assert!(flags.has_flag("NOISE_REMOVAL_SKIPPED"));
//! ```

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

/// Individual quality flag with detailed context.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "flag_type")]
pub enum QualityFlag {
    /// NOISE-1: Thermal noise removal was skipped due to threshold checks.
    NoiseRemovalSkipped {
        reason: String,
        zero_ratio: Option<f64>,
        noise_power_ratio: Option<f64>,
        zero_threshold: f64,
        ratio_threshold: f64,
    },

    /// NOISE-1: Noise removal thresholds were modified by environment variables.
    NoiseThresholdsModified {
        zero_threshold: f64,
        ratio_threshold: f64,
        default_zero: f64,
        default_ratio: f64,
    },

    /// DEBURST-1: Polynomial t0 fallback occurred.
    DeburstT0Fallback {
        burst_id: usize,
        reason: String,
        expected_offset: Option<f64>,
    },

    /// DEBURST-1: Time domain mismatch detected in deburst.
    DeburstTimeDomainMismatch {
        burst_id: usize,
        burst_time: f64,
        poly_t0: f64,
        offset: f64,
    },

    /// RTC-1: Default incidence angle was used instead of annotation value.
    RtcDefaultIncidenceAngle {
        default_angle_deg: f64,
        reason: String,
    },

    /// RTC: Incidence angle profile was missing or incomplete.
    RtcIncidenceProfileMissing { reason: String },

    /// Calibration LUT parsing issue.
    CalibrationLutWarning { lut_type: String, issue: String },

    /// Generic warning flag for other conditions.
    Warning {
        code: String,
        message: String,
        details: HashMap<String, String>,
    },
}

impl QualityFlag {
    /// Get the flag type name for lookup.
    pub fn flag_type(&self) -> &'static str {
        match self {
            QualityFlag::NoiseRemovalSkipped { .. } => "NOISE_REMOVAL_SKIPPED",
            QualityFlag::NoiseThresholdsModified { .. } => "NOISE_THRESHOLDS_MODIFIED",
            QualityFlag::DeburstT0Fallback { .. } => "DEBURST_T0_FALLBACK",
            QualityFlag::DeburstTimeDomainMismatch { .. } => "DEBURST_TIME_DOMAIN_MISMATCH",
            QualityFlag::RtcDefaultIncidenceAngle { .. } => "RTC_DEFAULT_INCIDENCE_ANGLE",
            QualityFlag::RtcIncidenceProfileMissing { .. } => "RTC_INCIDENCE_PROFILE_MISSING",
            QualityFlag::CalibrationLutWarning { .. } => "CALIBRATION_LUT_WARNING",
            QualityFlag::Warning {  .. } => {
                // This is a bit awkward but we need a static str
                // In practice, use the code field for lookup
                "WARNING"
            }
        }
    }

    /// Get the severity level (for filtering/reporting).
    pub fn severity(&self) -> &'static str {
        match self {
            QualityFlag::NoiseRemovalSkipped { .. } => "WARNING",
            QualityFlag::NoiseThresholdsModified { .. } => "INFO",
            QualityFlag::DeburstT0Fallback { .. } => "WARNING",
            QualityFlag::DeburstTimeDomainMismatch { .. } => "ERROR",
            QualityFlag::RtcDefaultIncidenceAngle { .. } => "WARNING",
            QualityFlag::RtcIncidenceProfileMissing { .. } => "WARNING",
            QualityFlag::CalibrationLutWarning { .. } => "WARNING",
            QualityFlag::Warning { .. } => "WARNING",
        }
    }
}

/// Collection of quality flags for a processing run.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct QualityFlags {
    flags: Vec<QualityFlag>,
}

impl QualityFlags {
    /// Create an empty quality flags collection.
    pub fn new() -> Self {
        Self { flags: Vec::new() }
    }

    /// Add a quality flag.
    pub fn add(&mut self, flag: QualityFlag) {
        log::info!("🚩 Quality flag: {} - {:?}", flag.flag_type(), flag);
        self.flags.push(flag);
    }

    /// Check if a specific flag type is present.
    pub fn has_flag(&self, flag_type: &str) -> bool {
        self.flags.iter().any(|f| f.flag_type() == flag_type)
    }

    /// Get all flags.
    pub fn all_flags(&self) -> &[QualityFlag] {
        &self.flags
    }

    /// Get flags by severity.
    pub fn flags_by_severity(&self, severity: &str) -> Vec<&QualityFlag> {
        self.flags
            .iter()
            .filter(|f| f.severity() == severity)
            .collect()
    }

    /// Check if any error-level flags are present.
    pub fn has_errors(&self) -> bool {
        self.flags.iter().any(|f| f.severity() == "ERROR")
    }

    /// Check if any warning-level flags are present.
    pub fn has_warnings(&self) -> bool {
        self.flags.iter().any(|f| f.severity() == "WARNING")
    }

    /// Get count of flags.
    pub fn count(&self) -> usize {
        self.flags.len()
    }

    /// Merge flags from another collection.
    pub fn merge(&mut self, other: QualityFlags) {
        self.flags.extend(other.flags);
    }

    /// Convert to JSON string for output metadata.
    pub fn to_json(&self) -> String {
        serde_json::to_string_pretty(&self).unwrap_or_else(|_| "[]".to_string())
    }

    /// Convert to a summary map for embedding in metadata.
    pub fn to_summary(&self) -> HashMap<String, Vec<String>> {
        let mut summary: HashMap<String, Vec<String>> = HashMap::new();

        for flag in &self.flags {
            let key = flag.flag_type().to_string();
            let msg = match flag {
                QualityFlag::NoiseRemovalSkipped { reason, .. } => reason.clone(),
                QualityFlag::NoiseThresholdsModified {
                    zero_threshold,
                    ratio_threshold,
                    ..
                } => {
                    format!("zero={:.2}, ratio={:.2}", zero_threshold, ratio_threshold)
                }
                QualityFlag::DeburstT0Fallback {
                    burst_id, reason, ..
                } => {
                    format!("burst {}: {}", burst_id, reason)
                }
                QualityFlag::DeburstTimeDomainMismatch {
                    burst_id, offset, ..
                } => {
                    format!("burst {}: offset={:.1}s", burst_id, offset)
                }
                QualityFlag::RtcDefaultIncidenceAngle {
                    default_angle_deg,
                    reason,
                } => {
                    format!("{:.1}° used: {}", default_angle_deg, reason)
                }
                QualityFlag::RtcIncidenceProfileMissing { reason } => reason.clone(),
                QualityFlag::CalibrationLutWarning { lut_type, issue } => {
                    format!("{}: {}", lut_type, issue)
                }
                QualityFlag::Warning { code, message, .. } => {
                    format!("{}: {}", code, message)
                }
            };
            summary.entry(key).or_insert_with(Vec::new).push(msg);
        }

        summary
    }
}

/// Thread-safe global quality flag collector.
/// Used to accumulate flags across parallel processing stages.
#[derive(Clone)]
pub struct GlobalQualityFlags {
    inner: Arc<RwLock<QualityFlags>>,
}

impl GlobalQualityFlags {
    /// Create a new global quality flag collector.
    pub fn new() -> Self {
        Self {
            inner: Arc::new(RwLock::new(QualityFlags::new())),
        }
    }

    /// Add a flag (thread-safe).
    pub fn add(&self, flag: QualityFlag) {
        if let Ok(mut flags) = self.inner.write() {
            flags.add(flag);
        }
    }

    /// Get a copy of all flags.
    pub fn get_flags(&self) -> QualityFlags {
        self.inner.read().map(|f| f.clone()).unwrap_or_default()
    }

    /// Clear all flags.
    pub fn clear(&self) {
        if let Ok(mut flags) = self.inner.write() {
            flags.flags.clear();
        }
    }
}

impl Default for GlobalQualityFlags {
    fn default() -> Self {
        Self::new()
    }
}

// ============================================================================
// GLOBAL QUALITY FLAG SINGLETON
// ============================================================================

use std::sync::OnceLock;

static GLOBAL_QUALITY_FLAGS: OnceLock<GlobalQualityFlags> = OnceLock::new();

/// Get the global quality flag collector (singleton).
///
/// This allows processing functions (noise removal, deburst, RTC) to emit
/// quality flags without requiring API changes. The flags can be retrieved
/// at the end of processing for inclusion in output metadata.
///
/// # Example
///
/// ```rust,ignore
/// use sardine::core::quality_flags::{global_quality_flags, QualityFlag};
///
/// // During processing:
/// global_quality_flags().add(QualityFlag::NoiseRemovalSkipped { ... });
///
/// // At end of processing:
/// let flags = global_quality_flags().get_flags();
/// let json = flags.to_json();
/// ```
pub fn global_quality_flags() -> &'static GlobalQualityFlags {
    GLOBAL_QUALITY_FLAGS.get_or_init(GlobalQualityFlags::new)
}

/// Reset the global quality flags (for testing or between processing runs).
/// Note: This creates a new GlobalQualityFlags, the old one becomes inaccessible.
pub fn reset_global_quality_flags() {
    if let Some(flags) = GLOBAL_QUALITY_FLAGS.get() {
        flags.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_quality_flag_creation() {
        let flag = QualityFlag::NoiseRemovalSkipped {
            reason: "Test reason".to_string(),
            zero_ratio: Some(0.96),
            noise_power_ratio: Some(0.8),
            zero_threshold: 0.95,
            ratio_threshold: 0.9,
        };

        assert_eq!(flag.flag_type(), "NOISE_REMOVAL_SKIPPED");
        assert_eq!(flag.severity(), "WARNING");
    }

    #[test]
    fn test_quality_flags_collection() {
        let mut flags = QualityFlags::new();

        flags.add(QualityFlag::DeburstT0Fallback {
            burst_id: 1,
            reason: "Missing t0".to_string(),
            expected_offset: None,
        });

        assert!(flags.has_flag("DEBURST_T0_FALLBACK"));
        assert!(!flags.has_flag("NOISE_REMOVAL_SKIPPED"));
        assert_eq!(flags.count(), 1);
    }

    #[test]
    fn test_quality_flags_json() {
        let mut flags = QualityFlags::new();
        flags.add(QualityFlag::RtcDefaultIncidenceAngle {
            default_angle_deg: 35.0,
            reason: "Annotation not available".to_string(),
        });

        let json = flags.to_json();
        // The flag_type is serialized as a tag by serde
        assert!(json.contains("RtcDefaultIncidenceAngle") || json.contains("flag_type"));
        assert!(json.contains("35.0") || json.contains("35"));
    }

    #[test]
    fn test_global_quality_flags_thread_safe() {
        let global = GlobalQualityFlags::new();

        let g1 = global.clone();
        let g2 = global.clone();

        std::thread::spawn(move || {
            g1.add(QualityFlag::Warning {
                code: "TEST1".to_string(),
                message: "Thread 1".to_string(),
                details: HashMap::new(),
            });
        })
        .join()
        .unwrap();

        std::thread::spawn(move || {
            g2.add(QualityFlag::Warning {
                code: "TEST2".to_string(),
                message: "Thread 2".to_string(),
                details: HashMap::new(),
            });
        })
        .join()
        .unwrap();

        let flags = global.get_flags();
        assert_eq!(flags.count(), 2);
    }
}
